src/wrapper.cpp

```diff
@@ -23,47 +28,73 @@ extern "C" {
 #endif
 
 #ifdef NO_CXXEXCEPTIONS
-#define THROW(expr) (last_exception = expr);
-#define CLEAR_LAST_EXCEPTION(expr) do { DCHECK(!exception_occurred()); last_exception = TokenException::none(); } while (0);
+#define THROW(expr) TokenException::set_last_exception(expr);
+#define CLEAR_LAST_EXCEPTION(expr) \
+    DCHECK(!exception_occurred()); \
+    TokenException::set_last_exception(TokenException::none());
```

Comment:
good catch, we dropped this when we switched from `thread_local` to `ThreadLocalPointer`

Reply:
@riastradh-brave why?

Reply:
in either case we should still keep this multi-line for readability

Reply:
a problem can arise when you do something like:
```
if (foo)
  CLEAR_LAST_EXCEPTION();
```
(contrived and obviously bad practice)

Reply:
I see, makes sense, but wouldn't a regular block work just as well? `{ ... }` without the do/while? It's fine to have nested blocks if it's already in a block, but wfm either way

Reply:
the problem with that arises if you have 
```
if (foo)
  CLEAR_LAST_EXCEPTION();
else
  bar()
```
https://www.geeksforgeeks.org/multiline-macros-in-c/

Reply:
I guess my personal preference is the block inside parens, but that's purely for aesthetic reasons :)

Reply:
`do { ... } while (0)` is the standard way to turn multiple statements into a single statement that always works reliably.  If `DCHECK` expanded into a statement that is not an expression, then `(DCHECK(...), set_last_exception(...))` wouldn't work.

Reply:
@riastradh-brave `({ .. })` also works

Reply:
(and `({ ... })` is a nonstandard extension, as far as I'm aware)

Reply:
it really doesn't matter to me, I was just curious because there are a lot of multiline macros (including chromium code) that don't have anything around them, but the reasoning makes sense because in some cases it wouldn't even get caught by the compiler. As an interesting aside, this snippet is from logging.h and has all three version ;)
```
#define WRAPPED_TRAP_SEQUENCE() TRAP_SEQUENCE()
#else
#define WRAPPED_TRAP_SEQUENCE() \
  do {                          \
    [] { TRAP_SEQUENCE(); }();  \
  } while (false)
#endif

#define IMMEDIATE_CRASH()    \
  ({                         \
    WRAPPED_TRAP_SEQUENCE(); \
    __builtin_unreachable(); \
  })

// CHECK dies with a fatal error if condition is not true.  It is *not*
// controlled by NDEBUG, so the check will be executed regardless of
// compilation mode.
//
// We make sure CHECK et al. always evaluates their arguments, as
// doing CHECK(FunctionWithSideEffect()) is a common idiom.

#if defined(OFFICIAL_BUILD) && defined(NDEBUG)

// Make all CHECK functions discard their log strings to reduce code bloat, and
// improve performance, for official release builds.
//
// This is not calling BreakDebugger since this is called frequently, and
// calling an out-of-line function instead of a noreturn inline macro prevents
// compiler optimizations.
#define CHECK(condition) \
  UNLIKELY(!(condition)) ? IMMEDIATE_CRASH() : EAT_STREAM_PARAMETERS

// PCHECK includes the system error code, which is useful for determining
// why the condition failed. In official builds, preserve only the error code
// message so that it is available in crash reports. The stringified
// condition and any additional stream parameters are dropped.
#define PCHECK(condition)                                  \
  LAZY_STREAM(PLOG_STREAM(FATAL), UNLIKELY(!(condition))); \
  EAT_STREAM_PARAMETERS
```